QtAcademyProject.cpp

```diff
@@ -12,22 +12,155 @@
   QObject(parent),
-  m_map(new Map(BasemapStyle::OsmStandard, this))
+  m_map(new Map(this)),
+  m_basemap(new Basemap(BasemapStyle::OsmStandard, this))
 {
+  m_map->setBasemap(m_basemap);
+  m_tempObject.reset(new QObject(this));
 }
 
 QtAcademyProject::~QtAcademyProject()
 {
 }
 
+void QtAcademyProject::toggleOffline(bool offline)
+{
+  if (offline)
+  {
+    // Set the basemap to a null basemap
+    // Use a unique pointer so it will get cleaned up when there are no references to it
+    m_map->setBasemap(new Basemap(m_tempObject.get()));
+    // Load the offline base layers from memory
+    loadOfflineBasemaps();
+  }
+  else
+  {
+    // Set the map to the online basemap
+    m_map->setBasemap(m_basemap);
+
+    // Delete any offline objects by reseting the parent
+    m_tempObject.reset(new QObject(this));
+  }
+}
+
+void QtAcademyProject::loadOfflineBasemaps()
+{
+  // Check to see if any offline data exists
+  if (QFile::exists(vtpkPath+"/vectorTiles.vtpk"))
+  {
+    // Create a VectorTileCache object from the offline data
+    m_vectorTileCache = new VectorTileCache(vtpkPath + "/vectorTiles.vtpk", m_tempObject.get());
+
+    // Check to see if any additional item resources exist and use them in the constructor if so
+    if (QFile::exists(vtpkPath+"/itemResources"))
+    {
+      m_itemResourceCache = new ItemResourceCache(vtpkPath+"/itemResources", m_tempObject.get());
+      m_offlineLayer = new ArcGISVectorTiledLayer(m_vectorTileCache, m_itemResourceCache, m_tempObject.get());
+    }
+    else
+    {
+      m_offlineLayer = new ArcGISVectorTiledLayer(m_vectorTileCache, m_tempObject.get());
+    }
+
+    // Add the created ArcGISVectorTileLayer to the base layers
+    m_map->basemap()->baseLayers()->append(m_offlineLayer);
+  }
+}
+
+void QtAcademyProject::createOfflineAreaFromExtent()
+{
+  // Export all available ArcGISVectorTileLayers
+  for (Layer* layer : *(m_basemap->baseLayers()))
+  {
+    if (ArcGISVectorTiledLayer* vectorTileLayer = dynamic_cast<ArcGISVectorTiledLayer*>(layer))
+      exportVectorTiles(vectorTileLayer);
+  }
+}
+
+void QtAcademyProject::exportVectorTiles(ArcGISVectorTiledLayer* vectorTileLayer)
+{
+  // Create a new export task from the layer's source url
+  m_exportVectorTilesTask = new ExportVectorTilesTask(vectorTileLayer->url(), m_tempObject.get());
+
+  // Create default parameters for the layer service
+  // Normalize the central meridian in case the download area crosses the meridian
+  m_exportVectorTilesTask->createDefaultExportVectorTilesParametersAsync(GeometryEngine::normalizeCentralMeridian(m_mapView->visibleArea()), m_mapView->mapScale()*0.1)
+      .then(this, [this](ExportVectorTilesParameters defaultParams)
+  {
+    QDir path(vtpkPath);
+
+    // Remove any existing offline files
+    path.removeRecursively();
+
+    // Create the directory if it does not already exist
+    if (!QDir().exists(vtpkPath))
+      QDir().mkdir(vtpkPath);
```

Comment:
Is this redundant? You are calling `removeRecursively` on this path directly above.

Should it just be `path.mkdir(vtpkPath)`?